### PR TITLE
build v4.5.3 (with dynamic libraries)

### PR DESCRIPTION
**Pull request by grlee77**
I cleaned up conda-forge/staged-recipes#682 which builds shared SuiteSparse libraries compatible with the  open [scikit-umfpack PR](onda-forge/staged-recipes#683).  Sstatic libraries actually get built as well, but `make install` doesn't install them. 

I am now using SuiteSparse_config.mk as in master, but it no longer seems necessary to have separate files for linux and OS X.

I changed the license info in meta.yaml slightly.  I moved LGPL v2 first as that is what the majority of the code seems to be licensed under (I only found one library with a BSD 3-clause license).  In the v4.5.3 source the licenses are more clearly stated in the separate /Doc/License.txt within each library's subfolder.

Adding cmake to the build dependencies was needed for building the included Metis v5.1.0.

### CODE REVIEW

**Comment by jakirkham** [comment]
Is there some reason not to install both static and dynamic libraries?

**Comment by jakirkham** [comment]
> I am now using SuiteSparse_config.mk as in master, but it no longer seems necessary to have separate files for linux and OS X.

Yeah, I started to make them pretty similar at the end. For the most part, started using a bunch of `?=` instead of `=` allowed most of the configuration to happen in `build.sh`, which seemed preferable anyways. I emailed Tim Davis making this exact suggestion though I don't know if he will go for it.

**Comment by jakirkham** [comment]
> I changed the license info in meta.yaml slightly. I moved LGPL v2 first as that is what the majority of the code seems to be licensed under (I only found one library with a BSD 3-clause license). In the v4.5.3 source the licenses are more clearly stated in the separate /Doc/License.txt within each library's subfolder.

Yeah, that seems reasonable to me. Licensing here is a bit of a pain unfortunately. Appreciate that you added Apache 2.0 for METIS.

**Comment by jakirkham** [comment]
> Adding cmake to the build dependencies was needed for building the included Metis v5.1.0.

👍

File: recipe/SPQR_TBB.patch [hunk 1]
```diff
@@ -0,0 +1,11 @@
+--- SPQR/Lib/Makefile	2016-05-04 16:47:24.000000000 -0400
++++ SPQR/Lib/Makefile.new	2016-05-31 15:04:27.075922172 -0400
+@@ -13,7 +13,7 @@
+ include ../../SuiteSparse_config/SuiteSparse_config.mk
+ 
+ # SPQR depends on CHOLMOD, AMD, COLAMD, LAPACK, the BLAS and SuiteSparse_config
+-LDLIBS += -lamd -lcolamd -lcholmod -lsuitesparseconfig $(LAPACK) $(BLAS)
++LDLIBS += -lamd -lcolamd -lcholmod -lsuitesparseconfig $(LAPACK) $(BLAS) $(TBB)
```
Comment (jakirkham) [inline recipe/SPQR_TBB.patch]: Are we sure this patch is needed? It seemed to pick up TBB fine in 4.4.6 and 4.4.5. If this is indeed required, it might be good to drop Tim Davis a line about this regression.

File: recipe/build.sh [hunk 2]
```diff
@@ -19,5 +21,20 @@ export INSTALL_INCLUDE="${PREFIX}/include"
...
+# make install below fails to link libmetis unless it is copied to $PREFIX/lib
+cp $SRC_DIR/lib/libmetis$DYNAMIC_EXT $PREFIX/lib
+
 make install
+
+
+if [ "$(uname)" == "Darwin" ]; then
+
+  # change link to METIS for CHOLMOD to reflect install path
+  install_name_tool -change $SRC_DIR/lib/libmetis.dylib @rpath/libmetis.dylib $PREFIX/lib/libcholmod.dylib
```
Comment (jakirkham) [inline recipe/build.sh]: Why is this needed? Doesn't `conda-build` fix this anyways? If not, do we need to do something similar on Linux with `patchelf`?

File: recipe/build.sh [hunk 3]
```diff
@@ -19,5 +21,20 @@ export INSTALL_INCLUDE="${PREFIX}/include"
 export BLAS="-lopenblas"
 export LAPACK="-lopenblas"
 
+# (optional) write out various make variables for easier build debugging
+eval ${LIBRARY_SEARCH_VAR}="${PREFIX}/lib" make config 2>&1 | tee make_config.txt
+
 eval ${LIBRARY_SEARCH_VAR}="${PREFIX}/lib" make -j1
+
+# make install below fails to link libmetis unless it is copied to $PREFIX/lib
+cp $SRC_DIR/lib/libmetis$DYNAMIC_EXT $PREFIX/lib
```
Comment (jakirkham) [inline recipe/build.sh]: Maybe we could just build METIS first and install it. Then build everything else and install them. It might save us a few ugly hacks that cause us problems later.

File: recipe/SPQR_TBB.patch (hunk 1 above)
Comment (grlee77) [inline recipe/SPQR_TBB.patch]: I didn't need it on OS X, but for some reason when I tried locally on my linux machine I was getting errors related to TBB.  I can try removing it here and see what happens.

File: recipe/build.sh (hunk 2 above)
Comment (grlee77) [inline recipe/build.sh]: We can try removing it later.  I found that when I tried scikit-umfpack on OS X I got errors relating to the library not being found and this fixed it.

File: recipe/build.sh (hunk 3 above)
Comment (grlee77) [inline recipe/build.sh]: Do you mean as a separate feedstock or as a separate step in build.sh here?

**Comment by grlee77** [comment]
As far as the static libraries, they do also get built, but I didn't see a way to get `make install` to install them.  

After the make completes, I can see static libraries in various subfolders, e.g. for CHOLMOD the static lib is at `${SRC_DIR}/CHOLMOD/Lib/libcholmod.a`.  Obviously we could manually cp them, but maybe there is a way to do it from the Makefile that I overlooked?

File: recipe/build.sh (hunk 3 above)
Comment (jakirkham) [inline recipe/build.sh]: As part of `build.sh`.

Comment (grlee77) [inline recipe/build.sh]: Here is the relevant METIS section of the SuiteSparse Makefile
```
# configure and compile METIS, placing the libmetis.* library in
# SuiteSparse/lib and the metis.h include file in SuiteSparse/include.
metis: include/metis.h

# Install the shared version of METIS in SuiteSparse/lib.
# The SO_INSTALL_NAME commmand is only needed on the Mac, so *.dylib is
# hardcoded below.
include/metis.h:
ifeq (,$(MY_METIS_LIB))
	- ( cd metis-5.1.0 && $(MAKE) config shared=1 prefix=$(SUITESPARSE) )
	- ( cd metis-5.1.0 && $(MAKE) )
	- ( cd metis-5.1.0 && $(MAKE) install )
	- $(SO_INSTALL_NAME) $(SUITESPARSE)/lib/libmetis.dylib \
                             $(SUITESPARSE)/lib/libmetis.dylib
else
	@echo 'Using pre-installed METIS 5.1.0 library at ' '[$(MY_METIS_LIB)]'
endif
```
I think what we would want is prefix=$PREFIX so that it ends up installed where we would like as opposed to $SUITESPARSE ($SUITESPARSE is the same as $SRC_DIR)

Also, it seems like a bug in the Makefile that .dylib is listed there when the extension would be .so on linux?

**Comment by grlee77** [comment]
CI is now passing. will try removing the patch and adding install for the static libs

**Comment by jakirkham** [comment]
Seems there is an issue copying the static libs over. Here is part of the CircleCI [log]( https://circleci.com/gh/conda-forge/suitesparse-feedstock/10 ).

```
cp: cannot stat `/feedstock_root/build_artefacts/work/SuiteSparse/CHOLAMD/Lib/libcholamd.a': No such file or directory
Command failed: /bin/bash -x -e /recipe_root/build.sh
```

**Comment by grlee77** [comment]
I see there is a PR at staged-recipes for a METIS package as well.  How do we make sure the METIS compiled here doesn't end up conflicting with that one?

Apparently the SPQR_TBB.patch wasn't needed (not sure why I needed it to build successfully on my local linux system).

File: recipe/build.sh [hunk 4]
```diff
@@ -19,5 +21,46 @@ export INSTALL_INCLUDE="${PREFIX}/include"
...
+# build METIS first, installing into $PREFIX rather than the source DIR
+cd metis-5.1.0
+make config shared=1 prefix=${PREFIX}
+make
+make install
+
+if [ "$(uname)" == "Darwin" ]
+then
+    # this is copied from the SuiteSparse Makefile
+    install_name_tool -id ${PREFIX}/lib/libmetis.dylib ${PREFIX}/lib/libmetis.dylib
```
Comment (jakirkham) [inline recipe/build.sh]: Is this still needed now that we handle the build of METIS separately?

**Comment by jakirkham** [comment]
> I see there is a PR at staged-recipes for a METIS package as well. How do we make sure the METIS compiled here doesn't end up conflicting with that one?

I'm not sure how much we should push that one. My concern is the licensing with it is totally opaque. Also, we will want to move to SuiteSpare 4.5.x eventually, which includes a copy/version of METIS that has clear licensing. In any event, let's cross that bridge when we get there.

> Apparently the SPQR_TBB.patch wasn't needed (not sure why I needed it to build successfully on my local linux system).

Interesting. It does seem to be working. What version of Linux were you using?

<br>
Basically, this is looking really good. Thanks for doing this @grlee77. I'll try to find some time tomorrow or Friday to give it a closer look.

**Comment by grlee77** [comment]
I just noticed that the linux case is still passing `-fopenmp`.  I can fix that.

> Interesting. It does seem to be working. What version of Linux were you using?

Ubuntu 14.04.  I didn't look into it much, but the errors reported were:
```
spqr_parallel.o: In function `spqr_zippy<std::complex<double> >::~spqr_zippy()':
spqr_parallel.cpp:(.text._ZN10spqr_zippyISt7complexIdEED2Ev[_ZN10spqr_zippyISt7complexIdEED5Ev]+0x3): undefined reference to `vtable for tbb::task'
spqr_parallel.o: In function `spqr_zippy<double>::~spqr_zippy()':
spqr_parallel.cpp:(.text._ZN10spqr_zippyIdED2Ev[_ZN10spqr_zippyIdED5Ev]+0x3): undefined reference to `vtable for tbb::task'
spqr_parallel.o: In function `spqr_zippy<std::complex<double> >::~spqr_zippy()':
spqr_parallel.cpp:(.text._ZN10spqr_zippyISt7complexIdEED0Ev[_ZN10spqr_zippyISt7complexIdEED0Ev]+0x3): undefined reference to `vtable for tbb::task'
spqr_parallel.o: In function `spqr_zippy<double>::~spqr_zippy()':
spqr_parallel.cpp:(.text._ZN10spqr_zippyIdED0Ev[_ZN10spqr_zippyIdED0Ev]+0x3): undefined reference to `vtable for tbb::task'
spqr_parallel.o: In function `spqr_zippy<std::complex<double> >::execute()':
spqr_parallel.cpp:(.text._ZN10spqr_zippyISt7complexIdEE7executeEv[_ZN10spqr_zippyISt7complexIdEE7executeEv]+0x7e): undefined reference to `tbb::internal::allocate_child_proxy::allocate(unsigned long) const'
spqr_parallel.cpp:(.text._ZN10spqr_zippyISt7complexIdEE7executeEv[_ZN10spqr_zippyISt7complexIdEE7executeEv]+0xca): undefined reference to `tbb::task::spawn_and_wait_for_all(tbb::task_list&)'
spqr_parallel.o: In function `spqr_zippy<double>::execute()':
spqr_parallel.cpp:(.text._ZN10spqr_zippyIdE7executeEv[_ZN10spqr_zippyIdE7executeEv]+0x7e): undefined reference to `tbb::internal::allocate_child_proxy::allocate(unsigned long) const'
spqr_parallel.cpp:(.text._ZN10spqr_zippyIdE7executeEv[_ZN10spqr_zippyIdE7executeEv]+0xca): undefined reference to `tbb::task::spawn_and_wait_for_all(tbb::task_list&)'
spqr_parallel.o: In function `void spqr_parallel<double>(long, int, spqr_blob<double>*)':
spqr_parallel.cpp:(.text._Z13spqr_parallelIdEvliP9spqr_blobIT_E[_Z13spqr_parallelIdEvliP9spqr_blobIT_E]+0x2e): undefined reference to `tbb::task_scheduler_init::initialize(int, unsigned long)'
spqr_parallel.cpp:(.text._Z13spqr_parallelIdEvliP9spqr_blobIT_E[_Z13spqr_parallelIdEvliP9spqr_blobIT_E]+0x38): undefined reference to `tbb::internal::allocate_root_proxy::allocate(unsigned long)'
spqr_parallel.cpp:(.text._Z13spqr_parallelIdEvliP9spqr_blobIT_E[_Z13spqr_parallelIdEvliP9spqr_blobIT_E]+0x77): undefined reference to `tbb::task_scheduler_init::terminate()'
spqr_parallel.cpp:(.text._Z13spqr_parallelIdEvliP9spqr_blobIT_E[_Z13spqr_parallelIdEvliP9spqr_blobIT_E]+0x92): undefined reference to `tbb::task_scheduler_init::terminate()'
spqr_parallel.o: In function `void spqr_parallel<std::complex<double> >(long, int, spqr_blob<std::complex<double> >*)':
spqr_parallel.cpp:(.text._Z13spqr_parallelISt7complexIdEEvliP9spqr_blobIT_E[_Z13spqr_parallelISt7complexIdEEvliP9spqr_blobIT_E]+0x2e): undefined reference to `tbb::task_scheduler_init::initialize(int, unsigned long)'
spqr_parallel.cpp:(.text._Z13spqr_parallelISt7complexIdEEvliP9spqr_blobIT_E[_Z13spqr_parallelISt7complexIdEEvliP9spqr_blobIT_E]+0x38): undefined reference to `tbb::internal::allocate_root_proxy::allocate(unsigned long)'
spqr_parallel.cpp:(.text._Z13spqr_parallelISt7complexIdEEvliP9spqr_blobIT_E[_Z13spqr_parallelISt7complexIdEEvliP9spqr_blobIT_E]+0x77): undefined reference to `tbb::task_scheduler_init::terminate()'
spqr_parallel.cpp:(.text._Z13spqr_parallelISt7complexIdEEvliP9spqr_blobIT_E[_Z13spqr_parallelISt7complexIdEEvliP9spqr_blobIT_E]+0x92): undefined reference to `tbb::task_scheduler_init::terminate()'
spqr_parallel.o:(.data.rel.ro._ZTI10spqr_zippyIdE[_ZTI10spqr_zippyIdE]+0x10): undefined reference to `typeinfo for tbb::task'
spqr_parallel.o:(.data.rel.ro._ZTI10spqr_zippyISt7complexIdEE[_ZTI10spqr_zippyISt7complexIdEE]+0x10): undefined reference to `typeinfo for tbb::task'
spqr_parallel.o:(.data.rel.ro._ZTV10spqr_zippyIdE[_ZTV10spqr_zippyIdE]+0x28): undefined reference to `tbb::task::note_affinity(unsigned short)'
spqr_parallel.o:(.data.rel.ro._ZTV10spqr_zippyISt7complexIdEE[_ZTV10spqr_zippyISt7complexIdEE]+0x28): undefined reference to `tbb::task::note_affinity(unsigned short)'
collect2: error: ld returned 1 exit status
```

File: recipe/SPQR_TBB.patch (hunk 1 above)
Comment (jakirkham) [inline recipe/SPQR_TBB.patch]: We're no longer using this patch, right? Maybe we should remove it.

File: recipe/build.sh (hunk 4 above)
Comment (grlee77) [inline recipe/build.sh]: not sure if the `install_name_tool` call is necessary.  I copied it verbatim from the SuiteSparse Makefile.  I can try without and see if maybe it gets handled for us by conda build.

**Comment by jakirkham** [comment]
Looks like the OpenMP fix is working.

Is it possible the wrong compiler was getting picked up on Ubuntu 14.04? There are complaints like undefined reference to `vtable`, which look a bit strange.

**Comment by grlee77** [comment]
I think the last two commits have addressed the remaining comments.

I can squash a bunch of the commits if you want a cleaner history here.

I see that you already pinned `cvxopt` to use v4.4.* so merging this shouldn't break anything.  Then we can test the `scikit-umfpack` PR over at staged-recipes.

**Comment by jakirkham** [comment]
LGTM. Thanks @grlee77. Let's put it through its paces. 🏇
